utils.go

```diff
@@ -0,0 +1,201 @@
+package test
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"os"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"bytes"
+	"encoding/json"
+	"github.com/fnproject/fn_go/client"
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	"testing"
+)
+
+const lBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+func Host() string {
+	apiURL := os.Getenv("FN_API_URL")
+	if apiURL == "" {
+		apiURL = "http://localhost:8080"
+	}
+
+	u, err := url.Parse(apiURL)
+	if err != nil {
+		log.Fatalln("Couldn't parse API URL:", err)
+	}
+	return u.Host
+}
+
+func APIClient() *client.Fn {
+	transport := httptransport.New(Host(), "/v1", []string{"http"})
+	if os.Getenv("FN_TOKEN") != "" {
+		transport.DefaultAuthentication = httptransport.BearerToken(os.Getenv("FN_TOKEN"))
+	}
+
+	// create the API client, with the transport
+	return client.New(transport, strfmt.Default)
+}
+
+var (
+	appsandroutes = make(map[string][]string)
+	approutesLock sync.Mutex
+)
+
+type SuiteSetup struct {
+	Context      context.Context
+	Client       *client.Fn
+	AppName      string
+	RoutePath    string
+	Image        string
+	RouteType    string
+	Format       string
+	Memory       uint64
+	RouteConfig  map[string]string
+	RouteHeaders map[string][]string
+	Cancel       context.CancelFunc
+}
+
+func RandStringBytes(n int) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = lBytes[rand.Intn(len(lBytes))]
+	}
+	return strings.ToLower(string(b))
+}
+
+func SetupDefaultSuite() *SuiteSetup {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	ss := &SuiteSetup{
+		Context:      ctx,
+		Client:       APIClient(),
+		AppName:      "fnintegrationtestapp" + RandStringBytes(10),
+		RoutePath:    "/fnintegrationtestroute" + RandStringBytes(10),
+		Image:        "fnproject/hello",
+		Format:       "default",
+		RouteType:    "async",
+		RouteConfig:  map[string]string{},
+		RouteHeaders: map[string][]string{},
+		Cancel:       cancel,
+		Memory:       uint64(256),
+	}
+
+	if Host() != "localhost:8080" {
+		_, ok := http.Get(fmt.Sprintf("http://%s/version", Host()))
+		if ok != nil {
+			panic("Cannot reach remote api for functions")
+		}
+	}
+
+	return ss
+}
+
+func Cleanup() {
+	ctx := context.Background()
```

Comment:
Would be useful to optionally keep the app (e.g. via env) so that I can get logs for failures.